cpp/src/plasma/client.cc

```diff
@@ -72,6 +72,27 @@ constexpr int64_t kThreadPoolSize = 8;
 constexpr int64_t kBytesInMB = 1 << 20;
 static std::vector<std::thread> threadpool_(kThreadPoolSize);
 
+/// A Buffer class that automatically releases the backing plasma object
+/// when it goes out of scope.
+class PlasmaBuffer : public Buffer {
+ public:
+  ~PlasmaBuffer();
+
+  PlasmaBuffer(PlasmaClient* client, const ObjectID& object_id,
+               const std::shared_ptr<Buffer>& buffer)
+      : Buffer(buffer, 0, buffer->size()), client_(client), object_id_(object_id) {
+    if (buffer->is_mutable()) {
+      is_mutable_ = true;
+    }
+  }
+
+ private:
+  PlasmaClient* client_;
+  ObjectID object_id_;
+};
+
+PlasmaBuffer::~PlasmaBuffer() { (void)client_->Release(object_id_); }
```

Comment:
we have ARROW_UNUSED for this: https://github.com/apache/arrow/blob/ecb7605a640dc34452dd696240a74df9a9f7f39b/cpp/src/arrow/util/macros.h#L28